File: apps/nongyu-go-track-server/internal/store/sqlite/metrics.go
package sqlite

import (
	"context"
	"database/sql"
	"fmt"
	"math"
	"sort"
)

type Metric struct {
	StatDate string
	Key      string
	Value    int64
}

type DimRow struct {
	DimKey      string
	DimValue    string
	MetricValue int64
}

type CrashRow struct {
	EventID      string
	UserID       sql.NullInt64
	StudentNo    sql.NullString
	EventName    string
	AppVersion   sql.NullString
	Platform     sql.NullString
	DeviceBrand  sql.NullString
	ClientTsMs   sql.NullInt64
	ReceivedAtMs int64
	StatDate     string
	PropsJSON    sql.NullString
}

func (s *Store) UpsertMetric(ctx context.Context, tx *sql.Tx, date, key string, value, nowMs int64) error {
	_, err := tx.ExecContext(ctx, `
INSERT INTO daily_metrics (stat_date, metric_key, metric_value, updated_at_ms)
VALUES (?, ?, ?, ?)
ON CONFLICT(stat_date, metric_key) DO UPDATE SET
  metric_value=excluded.metric_value,
  updated_at_ms=excluded.updated_at_ms`, date, key, value, nowMs)
	if err != nil {
		return fmt.Errorf("upsert metric %s: %w", key, err)
	}
	return nil
}

func (s *Store) BumpPeak(ctx context.Context, date string, sample, nowMs int64) error {
	return s.WithWriteTx(ctx, func(tx *sql.Tx) error {
		var current sql.NullInt64
		err := tx.QueryRowContext(ctx, `
SELECT metric_value FROM daily_metrics WHERE stat_date=? AND metric_key='online_peak'`, date).Scan(&current)
		if err != nil && err != sql.ErrNoRows {
			return err
		}
		val := sample
		if current.Valid && current.Int64 > val {
			val = current.Int64
		}
		return s.UpsertMetric(ctx, tx, date, "online_peak", val, nowMs)
	})
}

func (s *Store) ReplaceDims(ctx context.Context, tx *sql.Tx, date, metricKey string, nowMs int64, rows []DimRow) error {
	if _, err := tx.ExecContext(ctx, `
DELETE FROM daily_dims WHERE stat_date=? AND metric_key=?`, date, metricKey); err != nil {
		return err
	}
	for _, row := range rows {
		if _, err := tx.ExecContext(ctx, `
INSERT INTO daily_dims (stat_date, metric_key, dim_key, dim_value, metric_value, updated_at_ms)
VALUES (?, ?, ?, ?, ?, ?)`, date, metricKey, row.DimKey, row.DimValue, row.MetricValue, nowMs); err != nil {
			return fmt.Errorf("insert dim: %w", err)
		}
	}
	return nil
}

func (s *Store) GetMetricMap(ctx context.Context, date string) (map[string]int64, error) {
	rows, err := s.read.QueryContext(ctx, `
SELECT metric_key, metric_value FROM daily_metrics WHERE stat_date=?`, date)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := map[string]int64{}
	for rows.Next() {
		var k string
		var v int64
		if err := rows.Scan(&k, &v); err != nil {
			return nil, err
		}
		out[k] = v
	}
	return out, rows.Err()
}

func (s *Store) Trend(ctx context.Context, metric, from, to string) ([]Metric, error) {
	rows, err := s.read.QueryContext(ctx, `
SELECT stat_date, metric_key, metric_value FROM daily_metrics
WHERE metric_key=? AND stat_date>=? AND stat_date<=?
ORDER BY stat_date ASC`, metric, from, to)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := make([]Metric, 0)
	for rows.Next() {
		var m Metric
		if err := rows.Scan(&m.StatDate, &m.Key, &m.Value); err != nil {
			return nil, err
		}
		out = append(out, m)
	}
	return out, rows.Err()
}

func (s *Store) Dims(ctx context.Context, metric, date string, limit int) ([]DimRow, error) {
	rows, err := s.read.QueryContext(ctx, `
SELECT dim_key, dim_value, metric_value FROM daily_dims
WHERE stat_date=? AND metric_key=?
ORDER BY metric_value DESC, dim_value ASC
LIMIT ?`, date, metric, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := make([]DimRow, 0)
	for rows.Next() {
		var d DimRow
		if err := rows.Scan(&d.DimKey, &d.DimValue, &d.MetricValue); err != nil {
			return nil, err
		}
		out = append(out, d)
	}
	return out, rows.Err()
}

// LiveDims 从当日 events 实时汇总维度（今日 overview/大屏用；不依赖 daily_dims）。
func (s *Store) LiveDims(ctx context.Context, metric, date string, limit int) ([]DimRow, error) {
	if limit < 1 {
		limit = 50
	}
	q := s.ReadDB()
	var rows []DimRow
	var err error
	switch metric {
	case "screen_views":
		// 进入次数：enter 无 duration_ms，避免与 leave 双计
		rows, err = s.CountScreenEnters(ctx, q, date)
	case "screen_dwell_avg":
		rows, err = s.AvgScreenDwell(ctx, q, date)
	case "button_clicks":
		rows, err = s.CountByName(ctx, q, date, "button_click")
	case "perf_p50", "perf_p95":
		perf, perr := s.PerfDurations(ctx, q, date)
		if perr != nil {
			return nil, perr
		}
		p := 50.0
		if metric == "perf_p95" {
			p = 95.0
		}
		rows = make([]DimRow, 0, len(perf))
		for name, vals := range perf {
			sorted := append([]int64(nil), vals...)
			sort.Slice(sorted, func(i, k int) bool { return sorted[i] < sorted[k] })
			rows = append(rows, DimRow{DimKey: "name", DimValue: name, MetricValue: Percentile(sorted, p)})
		}
	default:
		return nil, fmt.Errorf("unsupported live dim metric %s", metric)
	}
	if err != nil {
		return nil, err
	}
	sort.Slice(rows, func(i, k int) bool {
		if rows[i].MetricValue == rows[k].MetricValue {
			return rows[i].DimValue < rows[k].DimValue
		}
		return rows[i].MetricValue > rows[k].MetricValue
	})
	if len(rows) > limit {
		rows = rows[:limit]
	}
	return rows, nil
}

// Percentile 对已排序样本取分位（与日聚合口径一致）。
func Percentile(sorted []int64, p float64) int64 {
	if len(sorted) == 0 {
		return 0
	}
	idx := int(math.Ceil(p/100*float64(len(sorted)))) - 1
	if idx < 0 {
		idx = 0
	}
	if idx >= len(sorted) {
		idx = len(sorted) - 1
	}
	return sorted[idx]
}

func (s *Store) ListCrashes(ctx context.Context, from, to string, offset, limit int) ([]CrashRow, int, error) {
	return s.listEventsByType(ctx, "crash", from, to, "", offset, limit)
}

// ListLlmProxyFails 列出平台 LLM 代理失败事件；errorCode 为空则不过滤 event_name。
func (s *Store) ListLlmProxyFails(ctx context.Context, from, to, errorCode string, offset, limit int) ([]CrashRow, int, error) {
	return s.listEventsByType(ctx, "llm_proxy_fail", from, to, errorCode, offset, limit)
}

func (s *Store) listEventsByType(
	ctx context.Context,
	eventType, from, to, eventName string,
	offset, limit int,
) ([]CrashRow, int, error) {
	countSQL := `SELECT COUNT(*) FROM events WHERE event_type=? AND stat_date>=? AND stat_date<=?`
	listSQL := `
SELECT event_id, user_id, student_no, event_name, app_version, platform, device_brand,
       client_ts_ms, received_at_ms, stat_date, props_json
FROM events
WHERE event_type=? AND stat_date>=? AND stat_date<=?`
	argsCount := []any{eventType, from, to}
	argsList := []any{eventType, from, to}
	if eventName != "" {
		countSQL += ` AND event_name=?`
		listSQL += ` AND event_name=?`
		argsCount = append(argsCount, eventName)
		argsList = append(argsList, eventName)
	}
	listSQL += ` ORDER BY received_at_ms DESC LIMIT ? OFFSET ?`
	argsList = append(argsList, limit, offset)

	var total int
	if err := s.read.QueryRowContext(ctx, countSQL, argsCount...).Scan(&total); err != nil {
		return nil, 0, err
	}
	rows, err := s.read.QueryContext(ctx, listSQL, argsList...)
	if err != nil {
		return nil, 0, err
	}
	defer rows.Close()
	out := make([]CrashRow, 0)
	for rows.Next() {
		var c CrashRow
		if err := rows.Scan(
			&c.EventID, &c.UserID, &c.StudentNo, &c.EventName, &c.AppVersion, &c.Platform,
			&c.DeviceBrand, &c.ClientTsMs, &c.ReceivedAtMs, &c.StatDate, &c.PropsJSON,
		); err != nil {
			return nil, 0, err
		}
		out = append(out, c)
	}
	return out, total, rows.Err()
}

func (s *Store) CountDistinctDAU(ctx context.Context, q dbQuerier, date string) (int64, error) {
	var n int64
	err := q.QueryRowContext(ctx, `
SELECT COUNT(DISTINCT user_id) FROM events
WHERE stat_date=? AND event_type='app_open' AND user_id IS NOT NULL`, date).Scan(&n)
	return n, err
}

func (s *Store) CountByType(ctx context.Context, q dbQuerier, date, eventType string) (int64, error) {
	var n int64
	err := q.QueryRowContext(ctx, `
SELECT COUNT(*) FROM events WHERE stat_date=? AND event_type=?`, date, eventType).Scan(&n)
	return n, err
}

func (s *Store) CountByName(ctx context.Context, q dbQuerier, date, eventType string) ([]DimRow, error) {
	rows, err := q.QueryContext(ctx, `
SELECT event_name, COUNT(*) FROM events
WHERE stat_date=? AND event_type=?
GROUP BY event_name`, date, eventType)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := make([]DimRow, 0)
	for rows.Next() {
		var d DimRow
		d.DimKey = "name"
		if err := rows.Scan(&d.DimValue, &d.MetricValue); err != nil {
			return nil, err
		}
		out = append(out, d)
	}
	return out, rows.Err()
}

// CountScreenEnters 按 pathname 统计页面进入次数（无 duration_ms 的 screen_view）。
func (s *Store) CountScreenEnters(ctx context.Context, q dbQuerier, date string) ([]DimRow, error) {
	rows, err := q.QueryContext(ctx, `
SELECT event_name, COUNT(*) FROM events
WHERE stat_date=? AND event_type='screen_view' AND duration_ms IS NULL
GROUP BY event_name`, date)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := make([]DimRow, 0)
	for rows.Next() {
		var d DimRow
		d.DimKey = "name"
		if err := rows.Scan(&d.DimValue, &d.MetricValue); err != nil {
			return nil, err
		}
		out = append(out, d)
	}
	return out, rows.Err()
}

// AvgScreenDwell 按 pathname 对 leave 的 duration_ms 求平均（整数毫秒）。
func (s *Store) AvgScreenDwell(ctx context.Context, q dbQuerier, date string) ([]DimRow, error) {
	rows, err := q.QueryContext(ctx, `
SELECT event_name, CAST(ROUND(AVG(duration_ms)) AS INTEGER) FROM events
WHERE stat_date=? AND event_type='screen_view' AND duration_ms IS NOT NULL
GROUP BY event_name`, date)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := make([]DimRow, 0)
	for rows.Next() {
		var d DimRow
		d.DimKey = "name"
		if err := rows.Scan(&d.DimValue, &d.MetricValue); err != nil {
			return nil, err
		}
		out = append(out, d)
	}
	return out, rows.Err()
}

func (s *Store) PerfDurations(ctx context.Context, q dbQuerier, date string) (map[string][]int64, error) {
	rows, err := q.QueryContext(ctx, `
SELECT event_name, duration_ms FROM events
WHERE stat_date=? AND event_type='perf' AND duration_ms IS NOT NULL`, date)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := map[string][]int64{}
	for rows.Next() {
		var name string
		var ms int64
		if err := rows.Scan(&name, &ms); err != nil {
			return nil, err
		}
		out[name] = append(out[name], ms)
	}
	return out, rows.Err()
}

func (s *Store) PurgeEvents(ctx context.Context, cutoffMs int64) (int64, error) {
	var deleted int64
	err := s.WithWriteTx(ctx, func(tx *sql.Tx) error {
		res, err := tx.ExecContext(ctx, `DELETE FROM events WHERE received_at_ms < ?`, cutoffMs)
		if err != nil {
			return err
		}
		deleted, err = res.RowsAffected()
		return err
	})
	if err != nil {
		return 0, err
	}
	_, _ = s.write.ExecContext(ctx, "PRAGMA optimize")
	return deleted, nil
}

func (s *Store) RecordJob(ctx context.Context, name, key, status string, nowMs int64, detail string) error {
	_, err := s.write.ExecContext(ctx, `
INSERT INTO meta_jobs (job_name, job_key, status, finished_at_ms, detail_json)
VALUES (?, ?, ?, ?, ?)
ON CONFLICT(job_name, job_key) DO UPDATE SET
  status=excluded.status,
  finished_at_ms=excluded.finished_at_ms,
  detail_json=excluded.detail_json`, name, key, status, nowMs, nullIfEmpty(detail))
	return err
}

func (s *Store) JobStatus(ctx context.Context, name, key string) (string, bool, error) {
	var status string
	err := s.read.QueryRowContext(ctx, `
SELECT status FROM meta_jobs WHERE job_name=? AND job_key=?`, name, key).Scan(&status)
	if err == sql.ErrNoRows {
		return "", false, nil
	}
	if err != nil {
		return "", false, err
	}
	return status, true, nil
}

type dbQuerier interface {
	QueryContext(ctx context.Context, query string, args ...any) (*sql.Rows, error)
	QueryRowContext(ctx context.Context, query string, args ...any) *sql.Row
}

File: apps/nongyu-node-track-server/src/store/sqlite/metrics.ts
import type { Store } from "./db.js";
import { nullIfEmpty } from "./db.js";

export type Metric = { statDate: string; key: string; value: number };
export type DimRow = { dimKey: string; dimValue: string; metricValue: number };

export type DimFilter = {
  platform?: string;
  namePrefix?: string;
  /** 默认日聚合/App 大屏排除官网事件 */
  excludePlatform?: string;
};

export type CrashRow = {
  eventId: string;
  userId: number | null;
  studentNo: string | null;
  eventName: string;
  appVersion: string | null;
  platform: string | null;
  deviceBrand: string | null;
  clientTsMs: number | null;
  receivedAtMs: number;
  statDate: string;
  propsJson: string | null;
};

export function upsertMetric(
  store: Store,
  date: string,
  key: string,
  value: number,
  nowMs: number,
): void {
  store.db
    .prepare(
      `INSERT INTO daily_metrics (stat_date, metric_key, metric_value, updated_at_ms)
VALUES (?, ?, ?, ?)
ON CONFLICT(stat_date, metric_key) DO UPDATE SET
  metric_value=excluded.metric_value,
  updated_at_ms=excluded.updated_at_ms`,
    )
    .run(date, key, value, nowMs);
}

export function bumpPeak(store: Store, date: string, sample: number, nowMs: number): void {
  store.withWriteTx(() => {
    const current = store.db
      .prepare(
        `SELECT metric_value FROM daily_metrics WHERE stat_date=? AND metric_key='online_peak'`,
      )
      .get(date) as { metric_value: number } | undefined;
    let val = sample;
    if (current && current.metric_value > val) val = current.metric_value;
    upsertMetric(store, date, "online_peak", val, nowMs);
  });
}

export function replaceDims(
  store: Store,
  date: string,
  metricKey: string,
  nowMs: number,
  rows: DimRow[],
): void {
  store.db
    .prepare(`DELETE FROM daily_dims WHERE stat_date=? AND metric_key=?`)
    .run(date, metricKey);
  const ins = store.db.prepare(
    `INSERT INTO daily_dims (stat_date, metric_key, dim_key, dim_value, metric_value, updated_at_ms)
VALUES (?, ?, ?, ?, ?, ?)`,
  );
  for (const row of rows) {
    ins.run(date, metricKey, row.dimKey, row.dimValue, row.metricValue, nowMs);
  }
}

export function getMetricMap(store: Store, date: string): Record<string, number> {
  const rows = store.db
    .prepare(`SELECT metric_key, metric_value FROM daily_metrics WHERE stat_date=?`)
    .all(date) as Array<{ metric_key: string; metric_value: number }>;
  const out: Record<string, number> = {};
  for (const r of rows) out[r.metric_key] = r.metric_value;
  return out;
}

export function trend(store: Store, metric: string, from: string, to: string): Metric[] {
  const rows = store.db
    .prepare(
      `SELECT stat_date, metric_key, metric_value FROM daily_metrics
WHERE metric_key=? AND stat_date>=? AND stat_date<=?
ORDER BY stat_date ASC`,
    )
    .all(metric, from, to) as Array<{
    stat_date: string;
    metric_key: string;
    metric_value: number;
  }>;
  return rows.map((r) => ({ statDate: r.stat_date, key: r.metric_key, value: r.metric_value }));
}

/** 从当日 events 实时汇总维度（今日 overview/大屏用）。带 platform/namePrefix 时亦用于历史日。 */
const APP_DIM_FILTER: DimFilter = { excludePlatform: "web" };

export function liveDims(
  store: Store,
  metric: string,
  date: string,
  limit: number,
  filter?: DimFilter,
): DimRow[] {
  let lim = limit;
  if (lim < 1) lim = 50;
  const effective =
    filter?.platform || filter?.namePrefix || filter?.excludePlatform ? filter : APP_DIM_FILTER;
  let rows: DimRow[];
  switch (metric) {
    case "screen_views":
      rows = countScreenEnters(store, date, effective);
      break;
    case "screen_dwell_avg":
      rows = avgScreenDwell(store, date, effective);
      break;
    case "button_clicks":
      rows = countByName(store, date, "button_click", effective);
      break;
    case "perf_p50":
    case "perf_p95": {
      const perf = perfDurations(store, date, effective);
      const p = metric === "perf_p95" ? 95 : 50;
      rows = [];
      for (const [name, vals] of perf) {
        const sorted = [...vals].sort((a, b) => a - b);
        rows.push({ dimKey: "name", dimValue: name, metricValue: percentile(sorted, p) });
      }
      break;
    }
    default:
      throw new Error(`unsupported live dim metric ${metric}`);
  }
  rows.sort((a, b) => {
    if (a.metricValue === b.metricValue) return a.dimValue < b.dimValue ? -1 : 1;
    return b.metricValue - a.metricValue;
  });
  return rows.length > lim ? rows.slice(0, lim) : rows;
}

export function dims(
  store: Store,
  metric: string,
  date: string,
  limit: number,
  filter?: DimFilter,
): DimRow[] {
  if (filter?.platform || filter?.namePrefix || filter?.excludePlatform) {
    return liveDims(store, metric, date, limit, filter);
  }
  const rows = store.db
    .prepare(
      `SELECT dim_key, dim_value, metric_value FROM daily_dims
WHERE stat_date=? AND metric_key=?
ORDER BY metric_value DESC, dim_value ASC
LIMIT ?`,
    )
    .all(date, metric, limit) as Array<{
    dim_key: string;
    dim_value: string;
    metric_value: number;
  }>;
  return rows.map((r) => ({
    dimKey: r.dim_key,
    dimValue: r.dim_value,
    metricValue: r.metric_value,
  }));
}

/** 对已排序样本取分位（与日聚合口径一致）。 */
export function percentile(sorted: number[], p: number): number {
  if (sorted.length === 0) return 0;
  let idx = Math.ceil((p / 100) * sorted.length) - 1;
  if (idx < 0) idx = 0;
  if (idx >= sorted.length) idx = sorted.length - 1;
  return sorted[idx]!;
}

export function countDistinctDAU(store: Store, date: string): number {
  const row = store.db
    .prepare(
      `SELECT COUNT(DISTINCT user_id) AS n FROM events
WHERE stat_date=? AND event_type='app_open' AND user_id IS NOT NULL`,
    )
    .get(date) as { n: number };
  return row.n;
}

export function countByType(
  store: Store,
  date: string,
  eventType: string,
  platform?: string,
): number {
  if (platform) {
    const row = store.db
      .prepare(`SELECT COUNT(*) AS n FROM events WHERE stat_date=? AND event_type=? AND platform=?`)
      .get(date, eventType, platform) as { n: number };
    return row.n;
  }
  const row = store.db
    .prepare(`SELECT COUNT(*) AS n FROM events WHERE stat_date=? AND event_type=?`)
    .get(date, eventType) as { n: number };
  return row.n;
}

function appendEventFilters(
  sql: string,
  args: unknown[],
  filter?: DimFilter,
): { sql: string; args: unknown[] } {
  let out = sql;
  const next = [...args];
  if (filter?.platform) {
    out += ` AND platform=?`;
    next.push(filter.platform);
  }
  if (filter?.excludePlatform) {
    out += ` AND (platform IS NULL OR platform!=?)`;
    next.push(filter.excludePlatform);
  }
  if (filter?.namePrefix) {
    out += ` AND event_name LIKE ?`;
    next.push(`${filter.namePrefix}%`);
  }
  return { sql: out, args: next };
}

export function countByName(
  store: Store,
  date: string,
  eventType: string,
  filter?: DimFilter,
): DimRow[] {
  const base = appendEventFilters(
    `SELECT event_name, COUNT(*) AS n FROM events
WHERE stat_date=? AND event_type=?`,
    [date, eventType],
    filter,
  );
  const rows = store.db.prepare(`${base.sql} GROUP BY event_name`).all(...base.args) as Array<{
    event_name: string;
    n: number;
  }>;
  return rows.map((r) => ({ dimKey: "name", dimValue: r.event_name, metricValue: r.n }));
}

export function countScreenEnters(store: Store, date: string, filter?: DimFilter): DimRow[] {
  const base = appendEventFilters(
    `SELECT event_name, COUNT(*) AS n FROM events
WHERE stat_date=? AND event_type='screen_view' AND duration_ms IS NULL`,
    [date],
    filter,
  );
  const rows = store.db.prepare(`${base.sql} GROUP BY event_name`).all(...base.args) as Array<{
    event_name: string;
    n: number;
  }>;
  return rows.map((r) => ({ dimKey: "name", dimValue: r.event_name, metricValue: r.n }));
}

export function avgScreenDwell(store: Store, date: string, filter?: DimFilter): DimRow[] {
  const base = appendEventFilters(
    `SELECT event_name, CAST(ROUND(AVG(duration_ms)) AS INTEGER) AS n FROM events
WHERE stat_date=? AND event_type='screen_view' AND duration_ms IS NOT NULL`,
    [date],
    filter,
  );
  const rows = store.db.prepare(`${base.sql} GROUP BY event_name`).all(...base.args) as Array<{
    event_name: string;
    n: number;
  }>;
  return rows.map((r) => ({ dimKey: "name", dimValue: r.event_name, metricValue: r.n }));
}

export function perfDurations(
  store: Store,
  date: string,
  filter?: DimFilter,
): Map<string, number[]> {
  const base = appendEventFilters(
    `SELECT event_name, duration_ms FROM events
WHERE stat_date=? AND event_type='perf' AND duration_ms IS NOT NULL`,
    [date],
    filter,
  );
  const rows = store.db.prepare(base.sql).all(...base.args) as Array<{
    event_name: string;
    duration_ms: number;
  }>;
  const out = new Map<string, number[]>();
  for (const r of rows) {
    const list = out.get(r.event_name) ?? [];
    list.push(r.duration_ms);
    out.set(r.event_name, list);
  }
  return out;
}

function listEventsByType(
  store: Store,
  eventType: string,
  from: string,
  to: string,
  eventName: string,
  offset: number,
  limit: number,
): { rows: CrashRow[]; total: number } {
  let countSQL = `SELECT COUNT(*) AS n FROM events WHERE event_type=? AND stat_date>=? AND stat_date<=?`;
  let listSQL = `SELECT event_id, user_id, student_no, event_name, app_version, platform, device_brand,
       client_ts_ms, received_at_ms, stat_date, props_json
FROM events
WHERE event_type=? AND stat_date>=? AND stat_date<=?`;
  const argsCount: unknown[] = [eventType, from, to];
  const argsList: unknown[] = [eventType, from, to];
  if (eventName) {
    countSQL += ` AND event_name=?`;
    listSQL += ` AND event_name=?`;
    argsCount.push(eventName);
    argsList.push(eventName);
  }
  listSQL += ` ORDER BY received_at_ms DESC LIMIT ? OFFSET ?`;
  argsList.push(limit, offset);

  const total = (store.db.prepare(countSQL).get(...argsCount) as { n: number }).n;
  const raw = store.db.prepare(listSQL).all(...argsList) as Array<{
    event_id: string;
    user_id: number | null;
    student_no: string | null;
    event_name: string;
    app_version: string | null;
    platform: string | null;
    device_brand: string | null;
    client_ts_ms: number | null;
    received_at_ms: number;
    stat_date: string;
    props_json: string | null;
  }>;
  const rows = raw.map((c) => ({
    eventId: c.event_id,
    userId: c.user_id,
    studentNo: c.student_no,
    eventName: c.event_name,
    appVersion: c.app_version,
    platform: c.platform,
    deviceBrand: c.device_brand,
    clientTsMs: c.client_ts_ms,
    receivedAtMs: c.received_at_ms,
    statDate: c.stat_date,
    propsJson: c.props_json,
  }));
  return { rows, total };
}

export function listCrashes(store: Store, from: string, to: string, offset: number, limit: number) {
  return listEventsByType(store, "crash", from, to, "", offset, limit);
}

export function listLlmProxyFails(
  store: Store,
  from: string,
  to: string,
  errorCode: string,
  offset: number,
  limit: number,
) {
  return listEventsByType(store, "llm_proxy_fail", from, to, errorCode, offset, limit);
}

export function purgeEvents(store: Store, cutoffMs: number): number {
  let deleted = 0;
  store.withWriteTx(() => {
    const res = store.db.prepare(`DELETE FROM events WHERE received_at_ms < ?`).run(cutoffMs);
    deleted = res.changes;
  });
  try {
    store.db.exec("PRAGMA optimize");
  } catch {
    /* ignore */
  }
  return deleted;
}

export function recordJob(
  store: Store,
  name: string,
  key: string,
  status: string,
  nowMs: number,
  detail: string,
): void {
  store.db
    .prepare(
      `INSERT INTO meta_jobs (job_name, job_key, status, finished_at_ms, detail_json)
VALUES (?, ?, ?, ?, ?)
ON CONFLICT(job_name, job_key) DO UPDATE SET
  status=excluded.status,
  finished_at_ms=excluded.finished_at_ms,
  detail_json=excluded.detail_json`,
    )
    .run(name, key, status, nowMs, nullIfEmpty(detail));
}

export function jobStatus(
  store: Store,
  name: string,
  key: string,
): { status: string; ok: boolean } {
  const row = store.db
    .prepare(`SELECT status FROM meta_jobs WHERE job_name=? AND job_key=?`)
    .get(name, key) as { status: string } | undefined;
  if (!row) return { status: "", ok: false };
  return { status: row.status, ok: true };
}
